### PR TITLE
Add support ICP10125 pressure and temperature sensor

### PR DESCRIPTION
**Pull request by MizukiAgawa**
Product Homepage:
https://invensense.tdk.com/products/1-axis/icp-10125/

DataSheet:
https://3cfeqx1hf82y3xcoull08ihx-wpengine.netdna-ssl.com/wp-content/uploads/2021/04/DS-000329-ICP-10125-v1.1.pdf

Testing Environment:
STMicroelectronics NUCLEO-F401RE

Signed-off-by: Mizuki Agawa <agawa.mizuki@fujitsu.com>

### CODE REVIEW

File: drivers/sensor/icp10125/icp10125.c [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+	float A, B, C;
+
+	t = data->raw_ambient_temp - 32768.f;
+	in[0] = LUT_lower + (data->sensor_constants[0] * t * t) * quadr_factor;
+	in[1] = offst_factor * data->sensor_constants[3] +
+		(data->sensor_constants[1] * t * t) * quadr_factor;
+	in[2] = LUT_upper + (data->sensor_constants[2] * t * t) * quadr_factor;
+	icp10125_calculate_conversion_constants(in, &A, &B, &C);
+
+	return A + B / (C + data->raw_press);
```
Comment (gmarull) [inline drivers/sensor/icp10125/icp10125.c:121]: ditto

**Review by gmarull — DISMISSED** [review]
Sensor needs to be added to the build_all test for sensors as well

**Review by mbolivar-nordic — COMMENTED** [review]
ACK on the DT binding

File: drivers/sensor/icp10125/icp10125.c [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/byteorder.h>
+
+#ifdef CONFIG_ICP10125_CHECK_CRC
+#include <zephyr/sys/crc.h>
+#endif /* CONFIG_ICP10125_CHECK_CRC */
+
+LOG_MODULE_REGISTER(ICP10125, CONFIG_SENSOR_LOG_LEVEL);
+
+#define DT_DRV_COMPAT invensense_icp10125
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: Fixed it.

File: drivers/sensor/icp10125/icp10125.c [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+	struct icp10125_data *data = dev->data;
+	const struct icp10125_dev_config *cfg = dev->config;
+	uint8_t buffer[COMMAND_SIZE + OTP_READ_SETUP_PARAM_SIZE] = {
+		otp_read_setup_cmd.data[0], otp_read_setup_cmd.data[1], otp_read_setup_param[0],
+		otp_read_setup_param[1], otp_read_setup_param[2]};
+
+	if (i2c_write_dt(&cfg->i2c, buffer, COMMAND_SIZE + OTP_READ_SETUP_PARAM_SIZE)) {
+		LOG_ERR("Failed to write otp_read_setup.\n");
+		return -EIO;
+	}
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: Fixed it.

File: drivers/sensor/icp10125/icp10125.c [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+	return 0;
+}
+
+static int icp10125_check_crc(uint8_t *data)
+{
+	/* Details of CRC are described in Chapter 5 Section 8 of the product
+	 * specifications.
+	 */
+	return crc8(data, 2, ICP10125_CRC_POLY, 0xFF, false);
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: Fixed it.

File: drivers/sensor/icp10125/icp10125.c [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+	.sample_fetch = icp10125_sample_fetch,
+	.channel_get = icp10125_channel_get,
+};
+
+#define ICP10125_DEFINE(inst)                                                                      \
+	static struct icp10125_data icp10125_drv_##inst;                                           \
+	static const struct icp10125_dev_config icp10125_config_##inst = {                         \
+		.i2c = I2C_DT_SPEC_INST_GET(inst),                                                 \
+		.ambient_temp_mode = DT_ENUM_IDX(DT_DRV_INST(inst), temperature_measurement_mode), \
+		.press_mode = DT_ENUM_IDX(DT_DRV_INST(inst), pressure_measurement_mode)};          \
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: Fixed it.

File: drivers/sensor/icp10125/Kconfig [hunk 6]
```diff
@@ -0,0 +1,19 @@
+# ICP10125 barometric pressure/temperature sensor configuration options
+
+# Copyright (c) 2022 Mizuki Agawa <agawa.mizuki@fujitsu.com>
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig ICP10125
+	bool "ICP10125 Barometric Pressure & Temperature Sensor"
+	help
+	  Enable driver for ICP10125 barometric pressure/temperature sensor.
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/Kconfig]: Fixed it.

File: drivers/sensor/icp10125/icp10125.c [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+	const float p_Pa[] = {45000.0, 80000.0, 105000.0};
+
+	*C = (p_LUT[0] * p_LUT[1] * (p_Pa[0] - p_Pa[1]) +
+	      p_LUT[1] * p_LUT[2] * (p_Pa[1] - p_Pa[2]) +
+	      p_LUT[2] * p_LUT[0] * (p_Pa[2] - p_Pa[0])) /
+	     (p_LUT[2] * (p_Pa[0] - p_Pa[1]) + p_LUT[0] * (p_Pa[1] - p_Pa[2]) +
+	      p_LUT[1] * (p_Pa[2] - p_Pa[0]));
+	*A = (p_Pa[0] * p_LUT[0] - p_Pa[1] * p_LUT[1] - (p_Pa[1] - p_Pa[0]) * (*C)) /
+	     (p_LUT[0] - p_LUT[1]);
+	*B = (p_Pa[0] - (*A)) * (p_LUT[0] + (*C));
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: The 5.11 SAMPLE CODE in the datasheet shows a code fragment as a sample.
The datasheet does not mention the meaning of these parameters and functions.
I added a comment to copy it and reflect it in the source code.
- icp10125_calculate_conversion_constants()
- icp10125_calc_caribrated_ambient_temp()
- icp10125_calc_caribrated_press()

File: drivers/sensor/icp10125/icp10125.c [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+static float icp10125_calc_caribrated_ambient_temp(struct icp10125_data *data)
+{
+	/* Since there is no detailed description in the ICP10125 product manual,
+	 * the calculation of the pressure implement as same as show in the sample
+	 * code.
+	 */
+	return -45.f + 175.f / 65536.f * data->raw_ambient_temp;
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: same as a above.

File: drivers/sensor/icp10125/icp10125.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: The fix was wrong, so I fixed it. thank you for mentions.

File: drivers/sensor/icp10125/icp10125.c [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+#define MEASUREMENT_DATA_SIZE		 3
+#define AMBIENT_TEMP_DATA_SIZE		 (MEASUREMENT_DATA_SIZE * AMBIENT_TEMP_DATA_NUM)
+#define PRESS_DATA_SIZE			 (MEASUREMENT_DATA_SIZE * PRESS_DATA_NUM)
+#define PRESS_AND_AMBIENT_TEMP_DATA_SIZE (MEASUREMENT_DATA_SIZE * PRESS_AND_AMBIENT_TEMP_DATA_NUM)
+
+#define COMMAND_SIZE		  2
+#define OTP_DATA_SIZE		  3
+#define OTP_READ_SETUP_PARAM_SIZE 3
+
+#define SENSOR_DATA16(data, data_index) sys_get_be16(&data[data_index * MEASUREMENT_DATA_SIZE])
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: Corrected to not use the "SENSOR_DATA16" wrapper.
Define "struct sensor_data" to clarify data semantics.

File: drivers/sensor/icp10125/icp10125.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: Change the function to take a argument to determine the data length.

File: drivers/sensor/icp10125/icp10125.c [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+	}
+
+	return 0;
+}
+
+static void icp10125_convert_press_value(struct icp10125_data *data, struct sensor_value *val)
+{
+	float press = icp10125_calc_caribrated_press(data);
+
+	/* To prevent rounding errors, sensor_value_from_double() is not used  here. */
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: It is my mistake. The `sensor_value_from_double` is correct. Use it.

File: drivers/sensor/Kconfig [hunk 11]
```diff
@@ -254,4 +254,6 @@ source "drivers/sensor/ite_vcmp_it8xxx2/Kconfig"
 
 source "drivers/sensor/pcnt_esp32/Kconfig"
 
+source "drivers/sensor/icp10125/Kconfig"
```
Comment (MaureenHelm) [inline drivers/sensor/Kconfig:257]: Please move this up to sort in alpha order (yes there are a few outliers at the end that need to be fixed in another PR)

File: drivers/sensor/icp10125/icp10125.c [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+	     (p_LUT[0] - p_LUT[1]);
+	*B = (p_Pa[0] - (*A)) * (p_LUT[0] + (*C));
+}
+
+static float icp10125_calc_caribrated_ambient_temp(const struct icp10125_data *data)
+{
+	return -45.f + 175.f / 65536.f * data->raw_ambient_temp;
+}
+
+static float icp10125_calc_caribrated_press(const struct icp10125_data *data)
```
Comment (MaureenHelm) [inline drivers/sensor/icp10125/icp10125.c:125]:
```suggestion
static float icp10125_calc_calibrated_press(const struct icp10125_data *data)
```

File: dts/bindings/sensor/invensense,icp10125.yaml [hunk 13]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2022, Mizuki Agawa  <agawa.mizuki@fujitsu.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: ICP10125 barometric pressure/temperature sensor
+
+compatible: "invensense,icp10125"
+
+include: i2c-device.yaml
```
Comment (MaureenHelm) [inline dts/bindings/sensor/invensense,icp10125.yaml:8]:
```suggestion
include: [sensor-device.yaml, i2c-device.yaml]
```

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thank you for the contribution!

Please add this sensor to `tests/drivers/build_all/sensor` to ensure it gets built in CI.

File: drivers/sensor/CMakeLists.txt [hunk 14]
```diff
@@ -109,6 +109,7 @@ add_subdirectory_ifdef(CONFIG_ADC_CMP_NPCX	nuvoton_adc_cmp_npcx)
 add_subdirectory_ifdef(CONFIG_TACH_IT8XXX2	ite_tach_it8xxx2)
 add_subdirectory_ifdef(CONFIG_VCMP_IT8XXX2	ite_vcmp_it8xxx2)
 add_subdirectory_ifdef(CONFIG_PCNT_ESP32	pcnt_esp32)
+add_subdirectory_ifdef(CONFIG_ICP10125		icp10125)
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/CMakeLists.txt]: fixed it.

File: drivers/sensor/Kconfig (hunk 11 above)
Comment (MizukiAgawa) [reply drivers/sensor/Kconfig]: fixed it.

File: drivers/sensor/icp10125/icp10125.c [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+	      p_LUT[1] * p_LUT[2] * (p_Pa[1] - p_Pa[2]) +
+	      p_LUT[2] * p_LUT[0] * (p_Pa[2] - p_Pa[0])) /
+	     (p_LUT[2] * (p_Pa[0] - p_Pa[1]) + p_LUT[0] * (p_Pa[1] - p_Pa[2]) +
+	      p_LUT[1] * (p_Pa[2] - p_Pa[0]));
+	*A = (p_Pa[0] * p_LUT[0] - p_Pa[1] * p_LUT[1] - (p_Pa[1] - p_Pa[0]) * (*C)) /
+	     (p_LUT[0] - p_LUT[1]);
+	*B = (p_Pa[0] - (*A)) * (p_LUT[0] + (*C));
+}
+
+static float icp10125_calc_caribrated_ambient_temp(const struct icp10125_data *data)
```
[the remark this answers is not in the document]
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: fixed it.

File: drivers/sensor/icp10125/icp10125.c (hunk 12 above)
Comment (MizukiAgawa) [reply drivers/sensor/icp10125/icp10125.c]: fixed it.

File: dts/bindings/sensor/invensense,icp10125.yaml (hunk 13 above)
Comment (MizukiAgawa) [reply dts/bindings/sensor/invensense,icp10125.yaml]: fixed it.

**Comment by MizukiAgawa** [comment]
@MaureenHelm 
> Thank you for the contribution!
> 
> Please add this sensor to `tests/drivers/build_all/sensor` to ensure it gets built in CI.
Thank you for reviewing.
I added test_i2c_icp10125 node to `tests/drivers/build_all/sensor/i2c.dtsi`.

**Review by mbolivar-nordic — COMMENTED** [review]
@MizukiAgawa thank you for your review request. I no longer maintain the devicetree bindings. I asked for review from @galak instead on this area, since he is the current maintainer.

**Review by MaureenHelm — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]
